### PR TITLE
Changing AMP selection for AMD GPUs

### PR DESCRIPTION
**Pull request by stevenireeves**
Changed kMinGPUArch to be Vega10 for now. 

This will GPUArch will become string as per ROCm 3.7 as per https://github.com/ROCm-Developer-Tools/HIP/pull/2045
we will need to change the AMP selection function as such.

### CODE REVIEW

**Comment by ROCm-Apps-Test** [comment]
Can one of the admins verify this patch?

**Comment by ROCm-Apps-Test** [comment]
Can one of the admins verify this patch?

**Comment by ROCm-Apps-Test** [comment]
Can one of the admins verify this patch?

**Review by mvermeulen — APPROVED** [review]
Will defer to Deven on more detailed technical review.

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.cc
Comment (deven-amd) [inline tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1178]: Does `GetDeviceGPUArch` work correctly for AMD GPUs? 

Also I think for AMD, even though the numbers are in increasing order (gfx803, gfx900, gfx906, etc), the recommendation is to not use number comparison as a means for testing whether a given feature is supported. Because it is possible that 900 and 908 support some feature but 906 does not! 
So the routines for doing hardware specific feature checks for AMD should be something like 

```
...
return HasEnhancedFP16ComputeSupport(GetDeviceGPUArch(virtual_placer_.get_device(node)));
...
bool HasEnhancedFP16ComputeSupport(std::pair<int, int> gpu_arch)  
// data-type for gpu_arch arg can be anything you want...we are not limited to using int pair
{
// this routine should return true for gpu_arch that has enhanced fp16 support, false otherwise
}
```

Comment (stevenireeves) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1178]: Ok. I'll look into a more sophisticated approach for this.

Comment (deven-amd) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1178]: > Because it is possible that 900 and 908 support some feature but 906 does not!

that was meant as an example...not specific to fp16 support. For enhanced fp16 support (i.e. where we have fp16 compute support in HW, I believe gfx908 has it, gfx900 does not have it, and I do know for sure about the rest)

@mvermeulen, is there some wiki-page and / or document which lists which HW features are present in which gfx arch?

Comment (stevenireeves) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1178]: @deven-amp Peng sent me a doc with the features. I misread what gfx900 had, see my latest commit. 906 actually has the FP16 it appears, whereas 900 has "Packed" FP16.

**Comment by stevenireeves** [comment]
jenkins: retest please

**Comment by stevenireeves** [comment]
jenkins: retest rocm please

**Comment by stevenireeves** [comment]
jenkins: retest rocm please

**Comment by stevenireeves** [comment]
@deven-amd Can you give me any extra thoughts on this PR?

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.h
Comment (deven-amd) [inline tensorflow/core/grappler/optimizers/auto_mixed_precision.h:31]: nit: get rid of this blank space to remove the unnecessary diff

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.cc
Comment (deven-amd) [inline tensorflow/core/grappler/optimizers/auto_mixed_precision.cc]: @mvermeulen , can we put "90a" in open source yet?

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.cc
Comment (deven-amd) [inline tensorflow/core/grappler/optimizers/auto_mixed_precision.cc]: perhaps expose this in the .h file, to eliminate the need to cut-n-paste codes from here in the `*_test.cc` file?

**Review by deven-amd — COMMENTED** [review]
nit: please squash all your commits into one before the merge

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.cc
Comment (stevenireeves) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision.cc]: I originally put it into auto_mixed_precision.h but then I was a getting multiple definitions compilation error in the CI. Did you mean the function HasEnhancedFP16ComputeSupport, the data FP16ComputeSupportedDevices, or both?

Comment (deven-amd) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision.cc]: just the declaration of `HasEnhancedFP16ComputeSupport` in the .h file
that should get rid of the multiple definitions compilation error as well

Comment (stevenireeves) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision.cc]: That's right, thanks!

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.cc
Comment (stevenireeves) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision.cc]: I have a number of merge commits with develop-upstream. Rebasing and squashing is a bit difficult with all of these. @deven-amd

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.h
Comment (deven-amd) [inline tensorflow/core/grappler/optimizers/auto_mixed_precision.h]: why do we need to have this declaration in the header file?

File: tensorflow/core/grappler/optimizers/auto_mixed_precision_test.cc
Comment (deven-amd) [inline tensorflow/core/grappler/optimizers/auto_mixed_precision_test.cc]: can we figure out a way to avoid the copy-paste of `GetDeviceGPUArch` ?  
G will not let us upstream this code as is.

Comment (stevenireeves) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision_test.cc]: It's actually a different function that goes through all devices. I can rename it to avoid confusion.

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.h
Comment (stevenireeves) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision.h]: I had it this way so I wouldn't have two definitions of the var in `auto_mixed_precision.cc` and `auto_mixed_precision_test.cc` . Although `kMinGPUArch` is defined both in `auto_mixed_precision.cc` and `auto_mixed_precision_test.cc`. So I can do it that way.

**Comment by stevenireeves** [comment]
@deven-amd OK to merge?

**Comment by stevenireeves** [comment]
jenkins: retest rocm-cpp please

**Pull request merged**
